internal/openfga/stores.go

```diff
@@ -0,0 +1,343 @@
+	defer span.End()
+
+	roles, err := s.ofga.ListObjects(ctx, ID, "can_view", "role")
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, err
+	}
+
+	return roles, nil
+}
+
+// ListAssignedRoles returns all the roles a specific "assignee"able resource (user, group#member, role#assignee) is linked to (using "assignee" OpenFGA relation)
+func (s *OpenFGAStore) ListAssignedRoles(ctx context.Context, assigneeID string) ([]string, error) {
+	ctx, span := s.tracer.Start(ctx, "openfga.OpenFGAStore.ListAssignedRoles")
+	defer span.End()
+
+	roles, err := s.ofga.ListObjects(ctx, assigneeID, ASSIGNEE_RELATION, "role")
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, err
+	}
+
+	return roles, nil
+}
+
+// ListAssignedGroups returns all the groups a specific user is memeber of (using "member" OpenFGA relation)
+func (s *OpenFGAStore) ListAssignedGroups(ctx context.Context, assigneeID string) ([]string, error) {
+	ctx, span := s.tracer.Start(ctx, "openfga.OpenFGAStore.ListAssignedGroups")
+	defer span.End()
+
+	groups, err := s.ofga.ListObjects(ctx, assigneeID, MEMBER_RELATION, "group")
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, err
+	}
+
+	return groups, nil
+}
+
+// AssignRoles assigns roles to an "assignee"able resource (user, group#member)
+func (s *OpenFGAStore) AssignRoles(ctx context.Context, assigneeID string, roleIDs ...string) error {
+	ctx, span := s.tracer.Start(ctx, "openfga.OpenFGAStore.AssignRoles")
+	defer span.End()
+
+	// preemptive check to verify if all roles to be assigned are accessible by the user
+	// needs to happen separately
+
+	rs := make([]Tuple, 0)
+
+	for _, roleID := range roleIDs {
+		rs = append(rs, *NewTuple(assigneeID, ASSIGNEE_RELATION, roleID))
+	}
+
+	err := s.ofga.WriteTuples(ctx, rs...)
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+// UnassignRoles drops roles from an "assignee"able resource (user, group#member)
+func (s *OpenFGAStore) UnassignRoles(ctx context.Context, assigneeID string, roleIDs ...string) error {
+	ctx, span := s.tracer.Start(ctx, "openfga.OpenFGAStore.UnassignRoles")
+	defer span.End()
+
+	// preemptive check to verify if all roles to be assigned are accessible by the user
+	// needs to happen separately
+
+	rs := make([]Tuple, 0)
+
+	for _, roleID := range roleIDs {
+		rs = append(rs, *NewTuple(assigneeID, ASSIGNEE_RELATION, roleID))
+	}
+
+	err := s.ofga.DeleteTuples(ctx, rs...)
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+// AssignGroups assigns groups to an "assignee"able resource (user, group#member)
+func (s *OpenFGAStore) AssignGroups(ctx context.Context, assigneeID string, groupIDs ...string) error {
+	ctx, span := s.tracer.Start(ctx, "openfga.OpenFGAStore.AssignGroups")
+	defer span.End()
+
+	// preemptive check to verify if all Groups to be assigned are accessible by the user
+	// needs to happen separately
```

Comment:
question: this seems like a `// TODO`, should have an issue to track this or do you plan on adding this check in this PR?

Reply:
https://github.com/canonical/identity-platform-admin-ui/issues/387